sql/core/src/main/scala/org/apache/spark/sql/Column.scala

```diff
@@ -39,10 +39,10 @@ private[sql] object Column {
 }
 
 /**
- * A [[Column]] where an [[Encoder]] has been given for the expected return type.
+ * A [[Column]] where an [[Encoder]] has been given for the expected input and return type.
  * @since 1.6.0
  */
-class TypedColumn[T](expr: Expression)(implicit val encoder: Encoder[T]) extends Column(expr)
+class TypedColumn[-T, U](expr: Expression, val encoder: Encoder[U]) extends Column(expr)
```

Comment:
I think this is required if we want the ability to have typesafety and good inference for some expressions, while still allowing you to mix in the SQL expressions that are typechecked by the analyzer.

If you get rid of `T` then I don't think you'd need extra annotations for something like `sum(_._2)` and it wouldn't be type checked.  If you get rid of the annotation then you won't be able to use `expr("sum(_2)")` without casting.

I'll add some docs to clarify.